bigtable/client/detail/bulk_mutator.cc

```diff
@@ -0,0 +1,130 @@
+
+#include "bigtable/client/detail/bulk_mutator.h"
+
+#include <numeric>
+
+#include <bigtable/client/rpc_retry_policy.h>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+namespace detail {
+
+namespace btproto = google::bigtable::v2;
+
+BulkMutator::BulkMutator(std::string const &table_name,
+                         IdempotentMutationPolicy &policy, BulkMutation &&mut) {
+  mutations.set_table_name(table_name);
+  pending_mutations.set_table_name(table_name);
+  mut.MoveTo(&pending_mutations);
+  pending_original_index.resize(pending_mutations.entries_size());
+  std::iota(pending_original_index.begin(), pending_original_index.end(), 0);
+  pending_is_idempotent.reserve(pending_mutations.entries_size());
+  for (auto const &e : pending_mutations.entries()) {
+    auto r = std::all_of(e.mutations().begin(), e.mutations().end(),
+                         [&policy](btproto::Mutation const &m) {
+                           return policy.is_idempotent(m);
+                         });
+    pending_is_idempotent.push_back(r);
+  }
+}
+
+grpc::Status BulkMutator::MakeOneRequest(btproto::Bigtable::StubInterface &stub,
+                                         grpc::ClientContext &client_context) {
+  PrepareForRequest();
+  auto stream = stub.MutateRows(&client_context, mutations);
+  btproto::MutateRowsResponse response;
+  while (stream->Read(&response)) {
+    ProcessResponse(response);
+  }
+  FinishRequest();
+  return stream->Finish();
+}
+
+void BulkMutator::PrepareForRequest() {
+  mutations.Swap(&pending_mutations);
+  original_index.swap(pending_original_index);
+  is_idempotent.swap(pending_is_idempotent);
+  has_mutation_result = std::vector<bool>(mutations.entries_size());
+  pending_mutations = {};
+  pending_original_index = {};
+  pending_is_idempotent = {};
+}
+
+void BulkMutator::ProcessResponse(
+    google::bigtable::v2::MutateRowsResponse &response) {
+  for (auto &entry : *response.mutable_entries()) {
+    auto index = entry.index();
+    if (index < 0 or has_mutation_result.size() <= std::size_t(index)) {
+      // TODO(#72) - decide how this is logged.
+      continue;
+    }
+    has_mutation_result[index] = true;
+    auto &status = entry.status();
+    auto code = static_cast<grpc::StatusCode>(status.code());
+    if (code == grpc::OK) {
+      continue;
+    }
+    auto &original = *mutations.mutable_entries(index);
+    if (IsRetryableStatusCode(code) and is_idempotent[index]) {
+      pending_mutations.add_entries()->Swap(&original);
+      pending_original_index.push_back(original_index[index]);
+      pending_is_idempotent.push_back(is_idempotent[index]);
+    } else {
+      failures.emplace_back(SingleRowMutation(std::move(original)),
+                            std::move(*entry.mutable_status()),
+                            original_index[index]);
+    }
+  }
+}
+
+void BulkMutator::FinishRequest() {
+  // ... if there are any mutations with unknown state, try again ...
```

Comment:
Drop the "...", capitalize, and end sentence with a period, please.

Reply:
Done.  I thought I had found all of them.